src/lib.rs

```diff
@@ -573,6 +578,70 @@ mod tests {
         unwrap!(j1.join());
     }
 
+    // Test for the case when we send an extra handshake introducing ourselves as a Client after we already
+    // introduced ourselves as a Node. This message should be just ignored and the peer type should not
+    // be changed.
+    #[test]
+    fn double_handshake_node() {
+        let (mut qp2p0, rx0) = new_random_qp2p_for_unit_test(false, Default::default());
+        let qp2p0_info = unwrap!(qp2p0.our_connection_info());
+
+        let (tx1, _rx1) = mpsc::channel();
+        let mut malicious_client = TestPeer::new(tx1);
+        unwrap!(malicious_client.activate(OurType::Node));
+        malicious_client.send(
+            qp2p0_info.clone().into(),
+            WireMsg::Handshake(Handshake::Client),
+        );
+
+        std::thread::sleep(Duration::from_millis(10));
+
+        match rx0.try_recv() {
+            Ok(Event::ConnectedTo {
+                peer: Peer::Node { .. },
+            }) => {}
+            r => panic!("Unexpected result {:?}", r),
+        }
+
+        // No more messages expected
+        match rx0.try_recv() {
+            Err(TryRecvError::Empty) => {}
+            r => panic!("Unexpected result {:?}", r),
+        }
+    }
+
+    // Test for the case when we send an extra handshake introducing ourselves as a Node after we already
+    // introduced ourselves as a Client. This message should be just ignored and the peer type should not
+    // be changed.
+    #[test]
+    fn double_handshake_client() {
+        let (mut qp2p0, rx0) = new_random_qp2p_for_unit_test(false, Default::default());
+        let qp2p0_info = unwrap!(qp2p0.our_connection_info());
+
+        let (tx1, _rx1) = mpsc::channel();
+        let mut malicious_client = TestPeer::new(tx1);
+        unwrap!(malicious_client.activate(OurType::Client));
+        malicious_client.send(
+            qp2p0_info.clone().into(),
+            WireMsg::Handshake(Handshake::Node { cert_der: vec![] }),
+        );
+
+        std::thread::sleep(Duration::from_millis(10));
```

Comment:
Instead of delays can we not just do `recv()` below (followed by try_recv again ofc)